sage/example-transition/src/rules/hero_jam.rs

```diff
@@ -1,91 +1,119 @@
+	asset: &AssetId,
+) -> Result<(), RuleError>
 where
-	AccountId: Member + Codec,
 	BlockNumber: BlockNumberT,
-	AssetHandler: AssetManager<
-			AccountId = AccountId,
-			AssetId = asset::AssetId,
-			Asset = asset::Asset<BlockNumber>,
-		> + AssetInspector<
-			AccountId = AccountId,
-			AssetId = asset::AssetId,
-			Asset = asset::Asset<BlockNumber>,
-		>,
-	ChainHandler: ChainInspector<BlockNumber = BlockNumber>,
+	Inspector: AssetInspector<AccountId = AccountId, AssetId = AssetId, Asset = Asset<BlockNumber>>,
+	Chain: ChainInspector<BlockNumber = BlockNumber>,
 {
-	type AccountId = AccountId;
-	type AssetId = asset::AssetId;
+	let hero_jam_asset = try_cast_to_hero_jam_asset::<_, _, Inspector>(asset)?;
 
-	fn verify(
-		&self,
-		account_id: &Self::AccountId,
-		asset_ids: &[Self::AssetId],
-	) -> Result<(), Error> {
-		let filter_result = match self {
-			Rule::AllAssetType(expected_asset_type) => asset_ids.iter().all(|asset_id| {
-				if let Ok(asset) = AssetHandler::get_asset(asset_id) {
-					match asset.asset_variant {
-						AssetVariant::HeroJam(hero_jam_asset) =>
-							&hero_jam_asset.asset_type == expected_asset_type,
-					}
-				} else {
-					false
-				}
-			}),
-			Rule::AllStateType(expected_state_type) => asset_ids.iter().all(|asset_id| {
-				if let Ok(asset) = AssetHandler::get_asset(asset_id) {
-					match asset.asset_variant {
-						AssetVariant::HeroJam(hero_jam_asset) =>
-							&hero_jam_asset.state_type == expected_state_type,
-					}
-				} else {
-					false
-				}
-			}),
-			Rule::CanStateChange(_) => asset_ids.iter().all(|asset_id| {
-				if let Ok(asset) = AssetHandler::get_asset(asset_id) {
-					match asset.asset_variant {
-						AssetVariant::HeroJam(hero_jam_asset) =>
-							hero_jam_asset.state_change_block_number <
-								ChainHandler::get_current_block_number(),
-					}
-				} else {
-					false
-				}
-			}),
-			Rule::AccountHasAssetOfType(expected_asset_type, _) =>
-				AssetHandler::iter_assets_from(account_id)
-					.filter(|asset| match asset.asset_variant {
-						AssetVariant::HeroJam(hero_jam_asset) =>
-							&hero_jam_asset.asset_type == expected_asset_type,
-					})
-					.count()
-					.eq(&1),
-		};
+	if hero_jam_asset.state_change_block_number < Chain::get_current_block_number() {
+		Ok(())
+	} else {
+		Err(RuleError::Other { error: ASSET_CANNOT_STATE_CHANGE })
+	}
+}
 
-		filter_result.then_some(()).ok_or(Error::FeeError)
+pub(crate) fn ensure_account_has_asset_of_type<AccountId, BlockNumber, Inspector>(
+	account_id: &AccountId,
+	asset_type: AssetType,
+) -> Result<(), RuleError>
+where
+	Inspector: AssetInspector<AccountId = AccountId, AssetId = AssetId, Asset = Asset<BlockNumber>>,
+{
+	if Inspector::iter_assets_from(account_id)
+		.filter(|asset_id| {
+			if let Ok(asset) = try_cast_to_hero_jam_asset::<_, _, Inspector>(asset_id) {
+				asset.asset_type == asset_type
+			} else {
+				false
+			}
+		})
+		.count()
+		.eq(&1)
+	{
```

Comment:
I am pretty sure that we can replace this whole thing with a `.first(predicate)` call.